### PR TITLE
Fix sharder stuck

### PR DESCRIPTION
**Pull request by peterlimg**
## Fixes

- Fix a hidden cause of MPT 'node not found' error that stuck the sharders. The MPT `FullNode` `ExtensionNode` does not include the version number when calculation the node hash key. Suppose we have a `FullNode` that was removed in round 100 (not removed from db yet, happens later in client state prune worker), then in round 200, we added a new full node that has the same value, children nodes, those two `FullNode` will have the same node hash key. So when the prune client state worker start to work, the `FullNode` will be removed from db, and we will see 'node not found' error if we try to get that node from the round 200. The fix is to include the version number when calculating the node hash, so the `FullNode` in round 200 will not have the same key as it in round 100.
- Fixed a bug in sharder sync process that stuck the chain moving. This happens when block process was waiting for LFB moving as it reached the LFB+ahead round, while finalize block process was waiting for new block processed and push to the finalize channel, therefore, a dead lock happened. Fixed this by pushing back the block that is ahead of LFB+5 to the process channel so that it has the chance to be processed later.
- Implemented the MPT state validation on starting of sharders, it would roll back from latest finalized block till the one that has validate full state. 

## Changes
- Use `dead_nodes` to do client state pruning process #1210 

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/chaincore/chain/entity.go [hunk 1]
```diff
@@ -167,6 +167,8 @@ type Chain struct {
 	magicBlockSaver              MagicBlockSaver
 
 	pruneStats *util.PruneStats
+	// channel to trigger the worker to start prunce process immediately
```
Comment (dabasov) [inline code/go/0chain.net/chaincore/chain/entity.go]: typo

File: code/go/0chain.net/chaincore/chain/protocol_round.go [hunk 2]
```diff
@@ -136,9 +136,9 @@ func (c *Chain) FinalizeRoundImpl(r round.RoundI) {
 	if !r.SetFinalizing() {
 		logging.Logger.Debug("finalize_round: already finalizing",
 			zap.Int64("round", r.GetRoundNumber()))
-		if node.Self.Type == node.NodeTypeSharder {
-			return
-		}
+		//if node.Self.Type == node.NodeTypeSharder {
```
Comment (dabasov) [inline code/go/0chain.net/chaincore/chain/protocol_round.go]: dead code, can't get why we returfo sn harder previously

File: code/go/0chain.net/core/common/rate_limter.go [hunk 3]
```diff
@@ -47,7 +48,11 @@ func UserRateLimit(handler ReqRespHandlerf) ReqRespHandlerf {
 		return Recover(handler)
 	}
 	return func(writer http.ResponseWriter, request *http.Request) {
-		tollbooth.LimitFuncHandler(userRateLimit.Limiter, Recover(handler)).ServeHTTP(writer, request)
+		if strings.HasPrefix(request.URL.Path, "/_") {
```
Comment (dabasov) [inline code/go/0chain.net/core/common/rate_limter.go]: is it for `_diagnostics` to bypass limiter?

**Comment by dabasov** [comment]
just found that we don't use MPT.UpdateVersion(). I wonder how do we update version of node then?

File: code/go/0chain.net/core/common/rate_limter.go (hunk 3 above)
Comment (peterlimg) [reply code/go/0chain.net/core/common/rate_limter.go]: Yes, was for bypassing the diagnostic page for debugging, can be removed.

File: code/go/0chain.net/chaincore/chain/protocol_round.go (hunk 2 above)
Comment (peterlimg) [reply code/go/0chain.net/chaincore/chain/protocol_round.go]: > just found that we don't use MPT.UpdateVersion(). I wonder how do we update version of node then?

We don't need to update version of node now. We just record all the deleted nodes' `(key, version)`  in `dead_nodes` in PNode when finalize a block. When enough dead nodes are collected or the prune period hits, we will check the `dead_nodes` and remove nodes that has version below specific number from db.

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**